File: backend/app/models/network.py
from pydantic import BaseModel
from typing import List, Optional
from enum import Enum


class NodeType(str, Enum):
    junction = "junction"
    reservoir = "reservoir"


class PipeStatus(str, Enum):
    open = "open"
    closed = "closed"


class PipeCategory(str, Enum):
    trunk = "trunk"
    distribution = "distribution"
    service = "service"
    valve = "valve"


class NodeModel(BaseModel):
    id: str
    type: NodeType
    lat: float
    lon: float
    elevation_m: float
    base_demand_lps: float = 0.0
    zone: Optional[str] = None
    depth_from_reservoir: int = -1
    upstream_pipe_ids: List[str] = []
    downstream_pipe_ids: List[str] = []


class ReservoirModel(BaseModel):
    id: str
    lat: float
    lon: float
    head_m: float


class PipeModel(BaseModel):
    id: str
    start_node: str
    end_node: str
    length_m: float
    diameter_mm: float
    roughness_hw: float
    status: PipeStatus = PipeStatus.open
    category: PipeCategory = PipeCategory.service
    upstream_node_id: Optional[str] = None
    downstream_node_id: Optional[str] = None
    road_name: Optional[str] = None


class ZoneBoundary(BaseModel):
    """Approximate polygon boundary for map overlay (list of [lat, lon] pairs)."""
    coordinates: List[List[float]]


class ZoneModel(BaseModel):
    id: str
    display_name: str
    node_ids: List[str]
    boundary: ZoneBoundary


class NetworkTopologyResponse(BaseModel):
    nodes: List[NodeModel]
    pipes: List[PipeModel]
    reservoir: ReservoirModel
    zones: List[ZoneModel]


class NetworkStatusResponse(BaseModel):
    closed_pipes: List[str]
    total_pipes: int
    open_pipes: int


class PipeOperationRequest(BaseModel):
    pipe_ids: List[str]


class PipeOperationResponse(BaseModel):
    success: List[str]
    not_found: List[str]
    already_in_state: List[str]

File: frontend/src/types/network.ts
export type NodeType = 'junction' | 'reservoir';
export type PipeStatus = 'open' | 'closed';
export type PipeCategory = 'trunk' | 'distribution' | 'service' | 'valve';

export interface NetworkNode {
  id: string;
  type: NodeType;
  lat: number;
  lon: number;
  elevation_m: number;
  base_demand_lps: number;
  zone?: string;
  depth_from_reservoir: number;
  upstream_pipe_ids: string[];
  downstream_pipe_ids: string[];
}

export interface NetworkReservoir {
  id: string;
  lat: number;
  lon: number;
  head_m: number;
}

export interface NetworkPipe {
  id: string;
  start_node: string;
  end_node: string;
  length_m: number;
  diameter_mm: number;
  roughness_hw: number;
  status: PipeStatus;
  category: PipeCategory;
  upstream_node_id?: string;
  downstream_node_id?: string;
  road_name?: string;
}

export interface ZoneBoundary {
  coordinates: [number, number][];
}

export interface NetworkZone {
  id: string;
  display_name: string;
  node_ids: string[];
  boundary: ZoneBoundary;
}

export interface NetworkTopology {
  nodes: NetworkNode[];
  pipes: NetworkPipe[];
  reservoir: NetworkReservoir;
  zones: NetworkZone[];
}

export interface NetworkStatus {
  closed_pipes: string[];
  total_pipes: number;
  open_pipes: number;
}
